src/v/cluster/scheduling/partition_allocator.cc

```diff
@@ -0,0 +1,207 @@
+// Copyright 2020 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "cluster/scheduling/partition_allocator.h"
+
+#include "cluster/fwd.h"
+#include "cluster/logger.h"
+#include "cluster/scheduling/constraints.h"
+#include "cluster/types.h"
+#include "model/metadata.h"
+
+#include <absl/container/node_hash_set.h>
+
+#include <algorithm>
+#include <exception>
+#include <iterator>
+#include <vector>
+
+namespace cluster {
+
+partition_allocator::partition_allocator()
+  : _state(std::make_unique<allocation_state>())
+  , _allocation_strategy(simple_allocation_strategy()) {}
+
+allocation_constraints default_constraints() {
+    allocation_constraints req;
+    req.hard_constraints.push_back(not_fully_allocated());
+    req.hard_constraints.push_back(not_decommissioned());
+    req.soft_constraints.push_back(least_allocated());
+    return req;
+}
+
+result<std::vector<model::broker_shard>>
+partition_allocator::allocate_partition(partition_constraints p_constraints) {
+    vlog(
+      clusterlog.trace,
+      "allocating partition with constraints: {}",
+      p_constraints);
+    if (
+      p_constraints.replication_factor <= 0
+      || _state->available_nodes() < p_constraints.replication_factor) {
+        return errc::topic_invalid_replication_factor;
+    }
+
+    intermediate_allocation<model::broker_shard> replicas(
+      *_state, p_constraints.replication_factor);
+
+    for (auto r = 0; r < p_constraints.replication_factor; ++r) {
+        auto effective_constraits = default_constraints();
+        effective_constraits.hard_constraints.push_back(
+          distinct_from(replicas.get()));
+        effective_constraits.add(std::move(p_constraints.constraints));
+
+        auto replica = _allocation_strategy.allocate_replica(
+          effective_constraits, *_state);
+
+        if (!replica) {
+            return replica.error();
+        }
+        replicas.push_back(replica.value());
+    }
+
+    return std::move(replicas).finish();
+}
+
+result<allocation_units>
+partition_allocator::allocate(allocation_request request) {
+    vlog(
+      clusterlog.trace,
+      "allocation request for {} partitions",
+      request.partitions.size());
+
+    intermediate_allocation<partition_assignment> assignments(
+      *_state, request.partitions.size());
+
+    for (auto& p_constraints : request.partitions) {
```

Comment:
From [the twitter exchange](https://twitter.com/ztlpn/status/1412574825684557830): I'm not sure that this is a problem in practice, but a simple one-by-one partition placement algorithm without backtracking can miss valid configurations that satisfy constraints. 

An example: say there are nodes A1, A2 in dc A, B1 in dc B, max 2 partitions per node. We need to place partitions P1, P2 and P3 with replication factor 2, P3 with replicas in different DCs. If we begin by placing P1 -> A1, B1; P2 -> A2, B1; P3 can't be placed, we need to backtrack.

Or, even simpler (without DC constraints): nodes N1, N2, N3, max 2 partitions per node, partitions P1, P2, P3 with replication factor 2. We begin by placing P1 -> N1, N2; P2 -> N1, N2; P3 can't be placed.

OTOH backtracking adds considerable complexity and results in exponential complexity 😅

Reply:
we should create a ticket and follow up w/ this work. 

@ztlpn would you want to hack on it?

Reply:
@senior7515 Possibly! I'd say this needs a bit more info about what kind of data and constraints are expected - proper heuristics can mean orders of magnitude difference in performance. Maybe even go the proper constraint solver route.

Reply:
@ztlpn - happy to chat async email (alex @ vectorized.io) or we can just migrate this to a github discussion. 

So in general our users want to tell the allocator where to place the replicas and preference for leaders (already implemented but not exposed). 

Think - 5 replicas:  2 datacenters, like so: `AZ1.DC1, AZ2.DC1, AZ3.DC1, AZ1.DC2, AZ2.DC2`, with preference to be in `DC1`, any `AZ`. 


in any case excited to continue the chat.

Reply:
@ztlpn great comments. Thank you for your contribution. We were also considering to go the proper constraint solver route. We will probably have to merge this first version in a shape similar to that what we have now. Currently usually operate in a space where we do not reach the limits (i.e. max number of partitions per node). Will we great to discuss further possibilities and solutions.